src/main/java/frc/robot/subsystems/swerve/Module.java

```diff
@@ -77,107 +56,59 @@ public void updateInputs() {
-
-      // Run closed loop drive control
-      // Only allowed if closed loop turn control is running
-      if (speedSetpoint != null) {
-        // Scale velocity based on turn error
-        //
-        // When the error is 90°, the velocity setpoint should be 0. As the wheel turns
-        // towards the setpoint, its velocity should increase. This is achieved by
-        // taking the component of the velocity in the direction of the setpoint.
-        double adjustSpeedSetpoint = speedSetpoint * Math.cos(turnFeedback.getPositionError());
-
-        // Run drive controller
-        double velocityRadPerSec = adjustSpeedSetpoint / WHEEL_RADIUS;
-        io.setDriveVoltage(
-            driveFeedforward.calculate(velocityRadPerSec)
-                + driveFeedback.calculate(inputs.driveVelocityRadPerSec, velocityRadPerSec));
-      }
-    }
+    Logger.processInputs(String.format("Swerve/%s Module", io.getModuleName()), inputs);
 
     // Calculate position deltas for odometry
-    int deltaCount =
-        Math.min(inputs.odometryDrivePositionsRad.length, inputs.odometryTurnPositions.length);
+    final int deltaCount =
+        Math.min(inputs.odometryDrivePositionsMeters.length, inputs.odometryTurnPositions.length);
     positionDeltas = new SwerveModulePosition[deltaCount];
     for (int i = 0; i < deltaCount; i++) {
-      double positionMeters = inputs.odometryDrivePositionsRad[i] * WHEEL_RADIUS;
-      Rotation2d angle =
-          inputs.odometryTurnPositions[i].plus(
-              turnRelativeOffset != null ? turnRelativeOffset : new Rotation2d());
+      final double positionMeters = inputs.odometryDrivePositionsMeters[i];
+      final Rotation2d angle = inputs.odometryTurnPositions[i];
       positionDeltas[i] = new SwerveModulePosition(positionMeters - lastPositionMeters, angle);
       lastPositionMeters = positionMeters;
     }
   }
 
-  /** Runs the module with the specified setpoint state. Returns the optimized state. */
+  /** Runs the module closed loop with the specified setpoint state. Returns the optimized state. */
   public SwerveModuleState runSetpoint(SwerveModuleState state) {
     // Optimize state based on current angle
-    // Controllers run in "periodic" when the setpoint is not null
-    var optimizedState = SwerveModuleState.optimize(state, getAngle());
+    final var optimizedState = SwerveModuleState.optimize(state, getAngle());
 
-    // Update setpoints, controllers run in "periodic"
-    angleSetpoint = optimizedState.angle;
-    speedSetpoint = optimizedState.speedMetersPerSecond;
+    io.setTurnSetpoint(optimizedState.angle);
+    io.setDriveSetpoint(optimizedState.speedMetersPerSecond);
 
     return optimizedState;
   }
 
   /** Runs the module with the specified voltage while controlling to zero degrees. */
   public void runCharacterization(double volts) {
     // Closed loop turn control
-    angleSetpoint = new Rotation2d();
+    io.setTurnSetpoint(new Rotation2d());
```

Comment:
You *might* want to make 0 degrees explicit if that's the behavior you expect here. Right now it relies on a default that could conceivably change.